File: 169.majority-element.py
#
# @lc app=leetcode id=169 lang=python3
#
# [169] Majority Element
#

# @lc code=start
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        count_map = {}
        for num in nums:
            if not num in count_map:
                count_map[num] = 0
            count_map[num] += 1
        for k,v in count_map.items():
            if v > len(nums)//2:
                return k       
# @lc code=end

File: golang/169.majority-element.go
/*
 * @lc app=leetcode id=169 lang=golang
 *
 * [169] Majority Element
 */
package main

import "fmt"

// @lc code=start
func majorityElement(nums []int) int {
	ret := nums[0]
	count := 1
	for _, num := range nums[1:] {
		if num == ret {
			count += 1
		} else {
			if count > 0 {
				count -= 1
			} else {
				count = 1
				ret = num

			}
		}
	}
	return ret
}

// @lc code=end
func main() {
	fmt.Println(majorityElement([]int{1, 3, 2, 2}))
	fmt.Println(majorityElement([]int{1, 3, 4, 2, 2}))
}
